ArduSub/mode.h

```diff
@@ -0,0 +1,679 @@
+                           float turn_rate_dps,
+                           int8_t direction,
+                           bool relative_angle);
+
+    private:
+
+        // yaw_cd(): main product of AutoYaw; the heading:
+        float yaw_cd();
+
+        // rate_cds(): desired yaw rate in centidegrees/second:
+        float rate_cds();
+
+        float look_ahead_yaw();
+        float roi_yaw();
+
+        // auto flight mode's yaw mode
+        uint8_t _mode = AUTO_YAW_LOOK_AT_NEXT_WP;
+
+        // Yaw will point at this location if mode is set to AUTO_YAW_ROI
+        Vector3f roi;
+
+        // bearing from current location to the ROI
+        float _roi_yaw;
+
+        // yaw used for YAW_FIXED yaw_mode
+        int32_t _fixed_yaw;
+
+        // Deg/s we should turn
+        int16_t _fixed_yaw_slewrate;
+
+        // heading when in yaw_look_ahead_yaw
+        float _look_ahead_yaw;
+
+        // used to reduce update rate to 100hz:
+        uint8_t roi_yaw_counter;
+
+        GuidedSubMode guided_mode;
+
+    };
+    static AutoYaw auto_yaw;
+
+    // pass-through functions to reduce code churn on conversion;
+    // these are candidates for moving into the Mode base
+    // class.
+    bool set_mode(Mode::Number mode, ModeReason reason);
+    GCS_Sub &gcs();
+
+    // end pass-through functions
+};
+
+class ModeManual : public Mode
+{
+
+public:
+    // inherit constructor
+    using Mode::Mode;
+
+    virtual void run() override;
+
+    bool init(bool ignore_checks) override;
+
+    bool requires_GPS() const override
+    {
+        return false;
+    }
+    bool has_manual_throttle() const override
+    {
+        return true;
+    }
+    bool allows_arming(bool from_gcs) const override
+    {
+        return true;
+    };
+    bool is_autopilot() const override
+    {
+        return false;
+    }
+
+protected:
+
+    const char *name() const override
+    {
+        return "MANUAL";
+    }
+    const char *name4() const override
+    {
+        return "MANU";
+    }
+
+private:
+
+};
+
+
+class ModeAcro : public Mode
+{
+
+public:
+    // inherit constructor
+    using Mode::Mode;
+
+    virtual void run() override;
+
+    bool init(bool ignore_checks) override;
+
+    bool requires_GPS() const override
+    {
+        return false;
+    }
+    bool has_manual_throttle() const override
+    {
+        return true;
+    }
+    bool allows_arming(bool from_gcs) const override
+    {
+        return true;
+    };
+    bool is_autopilot() const override
+    {
+        return false;
+    }
+
+protected:
+
+    const char *name() const override
+    {
+        return "ACRO";
+    }
+    const char *name4() const override
+    {
+        return "ACRO";
+    }
+
+private:
+
+};
+
+
+class ModeStabilize : public Mode
+{
+
+public:
+    // inherit constructor
+    using Mode::Mode;
+
+    virtual void run() override;
+
+    bool init(bool ignore_checks) override;
+
+    bool requires_GPS() const override
+    {
+        return false;
+    }
```

Comment:
```suggestion
    bool requires_GPS() const override { return false; }
```

... and elsewhere.  Very ugly in here otherwise :-)